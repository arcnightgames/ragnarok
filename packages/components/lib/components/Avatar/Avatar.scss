@use "sass:math";
@import "@arcnight/styles/scss/lib";

.avatar {
  &-container {
    --focus-outline: var(--norm-major-2);

    border: em(1);
    width: em(64);
    height: em(64);
    transition: 0.15s easing(ease-out-quint),
      background-position 0s;

    @supports selector(:focus-visible:has(a, b)) {
      // had to do this because of Safari
      outline: unset;
    }

    &,
    &:hover,
    &:focus,
    &:focus-within,
    &:active {
      text-decoration: none;
    }

    @supports not selector(:focus-visible:has(a, b)) {
      &:focus,
      &:focus-within {
        box-shadow: 0 0 0 #{$focus-ring-size} var(--focus-ring);
      }
    }

    @supports selector(:focus-visible:has(a, b)) {
      &:focus-visible,
      &:has(*:focus-visible) {
        box-shadow: 0 0 0 #{$focus-ring-size} var(--focus-ring);
      }
    }
  }

  &-container:has(.avatar-fallback) {
    background-color: var(--avatar-color);
  }

  &-img {
    width: em(64);
    height: em(64);
  }

  &-fallback {
    text-transform: uppercase;
    color: var(--primary-contrast);
    font-size: em(32);
  }
}

/*
 * Modifiers
 */
 .avatar-small {
  width: em(32);
  height: em(32);

  & .avatar-fallback {
    font-size: em(16);
  }
}

.avatar-large {
  width: em(128);
  height: em(128);

  & .avatar-fallback {
    font-size: em(64);
  }
}
