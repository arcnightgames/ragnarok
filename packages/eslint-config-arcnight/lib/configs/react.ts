import {
  pluginJSX,
  pluginReact,
  PluginReactHooks,
  pluginTestingLibrary,
} from "../plugins";
import {ConfigItem} from "../types";

export const react = (): ConfigItem[] => {
  return [
    {
      name: "arcnight:react",
      plugins: {
        jsx: pluginJSX,
        react: pluginReact,
        "react-hooks": PluginReactHooks,
        "testing-library": pluginTestingLibrary,
      },
    },
    {
      rules: {
        "no-underscore-dangle": [
          "error",
          {
            allow: ["__REDUX_DEVTOOLS_EXTENSION_COMPOSE__"],
            allowAfterThis: false,
            allowAfterSuper: false,
            enforceInMethodNames: true,
          },
        ],
        "class-methods-use-this": [
          "error",
          {
            exceptMethods: [
              "render",
              "getInitialState",
              "getDefaultProps",
              "getChildContext",
              "componentWillMount",
              "UNSAFE_componentWillMount",
              "componentDidMount",
              "componentWillReceiveProps",
              "UNSAFE_componentWillReceiveProps",
              "shouldComponentUpdate",
              "componentWillUpdate",
              "UNSAFE_componentWillUpdate",
              "componentDidUpdate",
              "componentWillUnmount",
              "componentDidCatch",
              "getSnapshotBeforeUpdate",
            ],
          },
        ],
        "react/display-name": ["off", {ignoreTranspilerName: false}],
        "react/forbid-prop-types": [
          "error",
          {
            forbid: ["any", "array", "object"],
            checkContextTypes: true,
            checkChildContextTypes: true,
          },
        ],
        "react/forbid-dom-props": ["off", {forbid: []}],
        "react/jsx-boolean-value": ["error", "never", {always: []}],
        "react/jsx-handler-names": [
          "off",
          {
            eventHandlerPrefix: "handle",
            eventHandlerPropPrefix: "on",
          },
        ],
        "react/jsx-key": "off",
        "react/jsx-no-bind": [
          "error",
          {
            ignoreRefs: true,
            allowArrowFunctions: true,
            allowFunctions: false,
            allowBind: false,
            ignoreDOMComponents: true,
          },
        ],
        "react/jsx-no-duplicate-props": ["error", {ignoreCase: true}],
        "react/jsx-no-literals": ["off", {noStrings: true}],
        "react/jsx-no-undef": "error",
        "react/jsx-pascal-case": [
          "error",
          {
            allowAllCaps: true,
            ignore: [],
          },
        ],
        "react/sort-prop-types": [
          "off",
          {
            ignoreCase: true,
            callbacksLast: false,
            requiredFirst: false,
            sortShapeProp: true,
          },
        ],
        "react/jsx-sort-prop-types": "off",
        "react/jsx-sort-props": [
          "off",
          {
            ignoreCase: true,
            callbacksLast: false,
            shorthandFirst: false,
            shorthandLast: false,
            noSortAlphabetically: false,
            reservedFirst: true,
          },
        ],
        "react/jsx-uses-react": ["error"],
        "react/jsx-uses-vars": "error",
        "react/no-danger": "warn",
        "react/no-deprecated": ["error"],
        "react/no-did-mount-set-state": "off",
        "react/no-did-update-set-state": "error",
        "react/no-will-update-set-state": "error",
        "react/no-direct-mutation-state": "off",
        "react/no-is-mounted": "error",
        "react/no-multi-comp": "off",
        "react/no-set-state": "off",
        "react/no-string-refs": "error",
        "react/no-unknown-property": "error",
        "react/prefer-es6-class": ["error", "always"],
        "react/prefer-stateless-function": [
          "error",
          {ignorePureComponents: true},
        ],
        "react/require-render-return": "error",
        "react/self-closing-comp": "error",
        "react/sort-comp": [
          "error",
          {
            order: [
              "static-variables",
              "static-methods",
              "instance-variables",
              "lifecycle",
              "/^handle.+$/",
              "/^on.+$/",
              "getters",
              "setters",
              "/^(get|set)(?!(InitialState$|DefaultProps$|ChildContext$)).+$/",
              "instance-methods",
              "everything-else",
              "rendering",
            ],
            groups: {
              lifecycle: [
                "displayName",
                "propTypes",
                "contextTypes",
                "childContextTypes",
                "mixins",
                "statics",
                "defaultProps",
                "constructor",
                "getDefaultProps",
                "getInitialState",
                "state",
                "getChildContext",
                "getDerivedStateFromProps",
                "componentWillMount",
                "UNSAFE_componentWillMount",
                "componentDidMount",
                "componentWillReceiveProps",
                "UNSAFE_componentWillReceiveProps",
                "shouldComponentUpdate",
                "componentWillUpdate",
                "UNSAFE_componentWillUpdate",
                "getSnapshotBeforeUpdate",
                "componentDidUpdate",
                "componentDidCatch",
                "componentWillUnmount",
              ],
              rendering: ["/^render.+$/", "render"],
            },
          },
        ],
        "react/jsx-no-target-blank": ["error", {enforceDynamicLinks: "always"}],
        "react/jsx-no-comment-textnodes": "error",
        "react/no-render-return-value": "error",
        "react/require-optimization": ["off", {allowDecorators: []}],
        "react/no-find-dom-node": "error",
        "react/forbid-elements": ["off", {forbid: []}],
        "react/no-danger-with-children": "error",
        "react/no-unused-prop-types": [
          "error",
          {
            customValidators: [],
            skipShapeProps: true,
          },
        ],
        "react/style-prop-object": "error",
        "react/no-unescaped-entities": "error",
        "react/no-children-prop": "error",
        "react/no-array-index-key": "error",
        "react/forbid-foreign-prop-types": ["warn", {allowInPropTypes: true}],
        "react/void-dom-elements-no-children": "error",
        "react/default-props-match-prop-types": [
          "error",
          {allowRequiredDefaults: false},
        ],
        "react/no-redundant-should-component-update": "error",
        "react/no-unused-state": "error",
        "react/boolean-prop-naming": [
          "off",
          {
            propTypeNames: ["bool", "mutuallyExclusiveTrueProps"],
            rule: "^(is|has)[A-Z]([A-Za-z0-9]?)+",
            message: "",
          },
        ],
        "react/no-typos": "error",
        "react/jsx-curly-brace-presence": [
          "error",
          {props: "never", children: "never"},
        ],
        "react/destructuring-assignment": ["error", "always"],
        "react/no-access-state-in-setstate": "error",
        "react/button-has-type": [
          "error",
          {
            button: true,
            submit: true,
            reset: false,
          },
        ],
        "react/jsx-child-element-spacing": "off",
        "react/no-this-in-sfc": "error",
        "react/jsx-max-depth": "off",
        "react/no-unsafe": "off",
        "react/jsx-fragments": ["error", "syntax"],
        "react/state-in-constructor": ["error", "always"],
        "react/static-property-placement": ["error", "property assignment"],
        "react/prefer-read-only-props": "off",
        "react/jsx-no-script-url": [
          "error",
          [
            {
              name: "Link",
              props: ["to"],
            },
          ],
        ],
        "react/jsx-no-useless-fragment": "error",
        "react/no-adjacent-inline-elements": "off",
        "react/function-component-definition": [
          "error",
          {
            namedComponents: ["arrow-function"],
          },
        ],
        "react/jsx-newline": "off",
        "react/jsx-no-constructed-context-values": "error",
        "react/no-unstable-nested-components": "error",
        "react/no-namespace": "error",
        "react/prefer-exact-props": "error",
        "react/no-arrow-function-lifecycle": "error",
        "react/no-invalid-html-attribute": "error",
        "react/no-unused-class-component-methods": "error",
        "react/hook-use-state": "off",
        "react/iframe-missing-sandbox": "off",
        "react/jsx-no-leaked-render": "off",
        "react/no-object-type-as-default-prop": "off",
        "react/sort-default-props": [
          "off",
          {
            ignoreCase: false,
          },
        ],
        "react/forbid-component-props": [
          "warn",
          {
            forbid: [
              {
                propName: "data-test-id",
                message: "Please use `data-testid` instead",
              },
            ],
          },
        ],
        "react/jsx-filename-extension": [1, {extensions: [".jsx", ".tsx"]}],
        "react/jsx-props-no-spreading": "off",
        "react/prop-types": "warn",
        "react/react-in-jsx-scope": "off",
        "react/require-default-props": "off",
        "react-hooks/rules-of-hooks": "error",
        "jsx/alt-text": "error",
        "jsx/anchor-has-content": "error",
        "jsx/aria-activedescendant-has-tabindex": "error",
        "jsx/aria-props": "error",
        "jsx/aria-proptypes": "error",
        "jsx/aria-role": "error",
        "jsx/aria-unsupported-elements": "error",
        "jsx/autocomplete-valid": "error",
        "jsx/heading-has-content": "error",
        "jsx/html-has-lang": "error",
        "jsx/iframe-has-title": "error",
        "jsx/no-access-key": "error",
        "jsx/no-distracting-elements": "error",
        "jsx/no-noninteractive-element-to-interactive-role": [
          "error",
          {
            ul: [
              "listbox",
              "menu",
              "menubar",
              "radiogroup",
              "tablist",
              "tree",
              "treegrid",
            ],
            ol: [
              "listbox",
              "menu",
              "menubar",
              "radiogroup",
              "tablist",
              "tree",
              "treegrid",
            ],
            li: ["menuitem", "option", "row", "tab", "treeitem"],
            table: ["grid"],
            td: ["gridcell"],
            fieldset: ["radiogroup", "presentation"],
          },
        ],
        "jsx/no-redundant-roles": "error",
        "jsx/role-has-required-aria-props": "error",
        "jsx/role-supports-aria-props": "error",
        "jsx/scope": "error",
        "jsx/tabindex-no-positive": "error",
        "jsx/anchor-ambiguous-text": "warn",
        "jsx/anchor-is-valid": "warn",
        "jsx/control-has-associated-label": "warn",
        "jsx/img-redundant-alt": "off",
        "jsx/interactive-supports-focus": "warn",
        "jsx/label-has-associated-control": "warn",
        "jsx/media-has-caption": "warn",
        "jsx/mouse-events-have-key-events": "warn",
        "jsx/no-aria-hidden-on-focusable": "warn",
        "jsx/no-autofocus": "off",
        "jsx/no-interactive-element-to-noninteractive-role": "warn",
        "jsx/no-noninteractive-tabindex": "warn",
        "jsx/no-static-element-interactions": "warn",
      },
      settings: {
        "import/resolver": {
          node: {
            extensions: [".js", ".jsx", ".json"],
          },
        },
        react: {
          version: "detect",
        },
        propWrapperFunctions: ["forbidExtraProps", "exact", "Object.freeze"],
      },
    },
  ];
};
