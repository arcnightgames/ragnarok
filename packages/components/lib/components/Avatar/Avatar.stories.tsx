import {Meta} from "@storybook/react";

import Avatar from "./Avatar";

export default {
  component: Avatar,
  title: "components/Avatar",
} as Meta<typeof Avatar>;

export const Playground = ({...args}) => <Avatar {...args} />;

Playground.argTypes = {
  size: {
    options: ["small", "medium", "large"],
    control: {
      type: "select",
      defaultValue: "medium",
    },
  },
  src: {
    control: "text",
  },
  fallback: {
    control: "text",
  },
};
