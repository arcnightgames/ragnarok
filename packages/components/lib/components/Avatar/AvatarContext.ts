import {noop} from "@arcnight/utils";
import {createContext} from "react";

export type ImageLoadingStatus = "idle" | "loading" | "loaded" | "error";

type AvatarContextInterface = {
  imageLoadingStatus: ImageLoadingStatus;
  onImageLoadingStatusChange(status: ImageLoadingStatus): void;
};

export default createContext<AvatarContextInterface>({
  imageLoadingStatus: "idle",
  onImageLoadingStatusChange: noop,
});
