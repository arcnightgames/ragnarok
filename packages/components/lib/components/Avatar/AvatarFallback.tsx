import "./Avatar.scss";

import {clsx} from "@arcnight/utils";
import {
  ForwardedRef,
  forwardRef,
  HTMLProps,
  useContext,
  useEffect,
  useState,
} from "react";

import AvatarContext from "./AvatarContext";

interface AvatarFallbackProps extends HTMLProps<HTMLSpanElement> {
  delayMs?: number;
  /**
   * Locator for e2e tests.
   */
  "data-testid"?: string;
}

const AvatarFallback = (
  {delayMs = 20, ...restProps}: AvatarFallbackProps,
  ref: ForwardedRef<HTMLSpanElement>,
) => {
  const {imageLoadingStatus} = useContext(AvatarContext);
  const [canRender, setCanRender] = useState(delayMs === undefined);

  useEffect(() => {
    if (delayMs) {
      const timerId = window.setTimeout(() => setCanRender(true), delayMs);
      return () => window.clearTimeout(timerId);
    }
  }, [delayMs]);

  return canRender && imageLoadingStatus !== "loaded" ? (
    <span
      className={clsx("avatar-fallback", "rounded-full")}
      {...restProps}
      ref={ref}
    />
  ) : null;
};

export default forwardRef(AvatarFallback);
