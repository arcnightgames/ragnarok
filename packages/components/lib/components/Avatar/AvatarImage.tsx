import "./Avatar.scss";

import {clsx} from "@arcnight/utils";
import {
  ForwardedRef,
  forwardRef,
  useContext,
  useLayoutEffect,
  useState,
} from "react";

import {AvatarSize} from "./Avatar";
import AvatarContext, {ImageLoadingStatus} from "./AvatarContext";

interface AvatarImageProps {
  /**
   * User avatar path
   */
  src: string;
  /**
   * Letter displayed when avatar loading error or missing avatar
   */
  alt?: string;
  /**
   * Controls how large the avatar should be.
   */
  size?: AvatarSize;
  /**
   * Locator for e2e tests.
   */
  "data-testid"?: string;
}

const useImageLoadingStatus = (src?: string) => {
  const [loadingStatus, setLoadingStatus] =
    useState<ImageLoadingStatus>("idle");

  useLayoutEffect(() => {
    if (!src) {
      setLoadingStatus("error");
      return;
    }

    let isMounted = true;
    const image = new window.Image();

    const updateStatus = (status: ImageLoadingStatus) => () => {
      if (!isMounted) {
        return;
      }

      setLoadingStatus(status);
    };

    setLoadingStatus("loading");
    image.onload = updateStatus("loaded");
    image.onerror = updateStatus("error");
    image.src = src;

    return () => {
      isMounted = false;
    };
  }, [src]);

  return loadingStatus;
};

const AvatarImage = (
  {src, alt, size = "medium", ...restProps}: AvatarImageProps,
  ref: ForwardedRef<HTMLImageElement>,
) => {
  const {onImageLoadingStatusChange} = useContext(AvatarContext);
  const imageLoadingStatus = useImageLoadingStatus(src);

  useLayoutEffect(() => {
    if (imageLoadingStatus !== "idle") {
      onImageLoadingStatusChange(imageLoadingStatus);
    }
  }, [imageLoadingStatus]);

  const avatarClassName = clsx(
    "avatar-img",
    "rounded-full",
    size !== "medium" && `avatar-${size}`,
  );

  return imageLoadingStatus === "loaded" ? (
    <img
      className={avatarClassName}
      {...restProps}
      ref={ref}
      src={src}
      alt={alt}
    />
  ) : null;
};

export default forwardRef(AvatarImage);
