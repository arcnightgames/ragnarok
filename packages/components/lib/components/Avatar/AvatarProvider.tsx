import "./Avatar.scss";

import {PolymorphicPropsWithRef} from "@arcnight/react-polymorphic-types";
import {clsx} from "@arcnight/utils";
import {ElementType, ForwardedRef, forwardRef, useMemo, useState} from "react";

import {AvatarSize} from "./Avatar";
import AvatarContext, {ImageLoadingStatus} from "./AvatarContext";

type AvatarProviderOwnProps = {
  /**
   * Controls how large the avatar should be.
   */
  size?: AvatarSize;
  /**
   * Locator for e2e tests.
   */
  "data-testid"?: string;
};

export type AvatarProviderProps<E extends ElementType> =
  PolymorphicPropsWithRef<AvatarProviderOwnProps, E>;

const defaultElement = "button";

const AvatarProvider = <E extends ElementType = typeof defaultElement>(
  {as, size = "medium", ...restProps}: AvatarProviderProps<E>,
  ref: ForwardedRef<Element>,
) => {
  const [imageLoadingStatus, setImageLoadingStatus] =
    useState<ImageLoadingStatus>("idle");

  const Element: ElementType = as || defaultElement;

  const avatarClassName = clsx(
    "avatar-container",
    "rounded-full",
    size !== "medium" && `avatar-${size}`,
  );

  return (
    <AvatarContext.Provider
      value={useMemo(
        () => ({
          imageLoadingStatus: imageLoadingStatus,
          onImageLoadingStatusChange: setImageLoadingStatus,
        }),
        [imageLoadingStatus],
      )}
    >
      <Element className={avatarClassName} ref={ref} {...restProps} />
    </AvatarContext.Provider>
  );
};

export default forwardRef(AvatarProvider);
