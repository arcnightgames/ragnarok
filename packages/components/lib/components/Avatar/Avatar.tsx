import "./Avatar.scss";

import {PolymorphicPropsWithRef} from "@arcnight/react-polymorphic-types";
import {ElementType, forwardRef, Ref} from "react";

import AvatarFallback from "./AvatarFallback";
import AvatarImage from "./AvatarImage";
import AvatarProvider from "./AvatarProvider";

export type AvatarSize = "small" | "medium" | "large";

interface AvatarOwnProps {
  /**
   * User avatar path
   */
  src: string;
  /**
   * Letter displayed when avatar loading error or missing avatar
   */
  fallback?: string;
  /**
   * Controls how large the avatar should be.
   */
  size?: AvatarSize;
  /**
   * Locator for e2e tests.
   */
  "data-testid"?: string;
}

export type AvatarProps<E extends ElementType> = PolymorphicPropsWithRef<
  AvatarOwnProps,
  E
>;

const defaultElement = "button";

const generateColor = (str?: string) => {
  if (!str) {
    return "var(--primary)";
  }

  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }

  hash = (hash ^ 0x1f0d88) & 0xffffff;

  const brightness =
    (0.2126 * ((hash >> 16) & 0xff) +
      0.7152 * ((hash >> 8) & 0xff) +
      0.0722 * (hash & 0xff)) /
    255;

  if (brightness < 0.2) {
    hash = hash + 0x666666;
  } else if (brightness > 0.8) {
    hash = hash - 0x666666;
  }

  let color = "#";
  for (let j = 0; j < 3; j++) {
    let value = (hash >> (j * 8)) & 0xff;
    color += ("00" + value.toString(16)).slice(-2);
  }

  return color;
};

const Avatar = <E extends ElementType = typeof defaultElement>(
  {
    src,
    fallback,
    size = "medium",
    as,
    "data-testid": dataTestId,
    ...restProps
  }: AvatarProps<E>,
  ref: Ref<Element>,
) => {
  return (
    <AvatarProvider
      {...restProps}
      style={{
        "--avatar-color": generateColor(fallback),
      }}
      size={size}
      ref={ref}
      as={as}
      data-testid={dataTestId && `${dataTestId}-provider`}
    >
      <AvatarImage
        alt={fallback}
        size={size}
        src={src}
        data-testid={dataTestId && `${dataTestId}-image`}
      />
      <AvatarFallback data-testid={dataTestId && `${dataTestId}-fallback`}>
        {fallback?.charAt(0)}
      </AvatarFallback>
    </AvatarProvider>
  );
};

export default forwardRef(Avatar);
