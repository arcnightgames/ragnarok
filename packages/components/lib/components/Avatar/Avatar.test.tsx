import {render, screen, waitFor} from "@testing-library/react";

import Avatar from "./Avatar";

// Mock the react-polymorphic-types module
vi.mock("@arcnight/react-polymorphic-types", () => ({
  PolymorphicPropsWithRef: vi.fn(),
}));

beforeAll(() => {
  global.Image = class {
    [x: string]: any;
    set src(src: string) {
      setTimeout(() => {
        if (src !== "https://cdn.arcnight.com/bad") {
          this.onload();
        } else {
          this.onerror();
        }
      }, 100);
    }
  } as any;
});

describe("Avatar Component", () => {
  it("renders with provided good src", async () => {
    const src = "https://cdn.arcnight.com/good";
    const fallback = "F";

    render(<Avatar src={src} fallback={fallback} data-testid="avatar" />);

    await waitFor(() => {
      const avatarImage = screen.getByAltText(fallback);
      expect(avatarImage).toBeInTheDocument();
      expect(avatarImage).toHaveAttribute("src", src);
    });
  });

  it("renders with provided bad src", async () => {
    const src = "https://cdn.arcnight.com/bad";
    const fallback = "F";

    render(<Avatar src={src} fallback={fallback} data-testid="avatar" />);

    await waitFor(() => {
      const avatarFallback = screen.getByTestId("avatar-fallback");
      expect(avatarFallback).toBeInTheDocument();
    });
  });

  it("renders with default size if size prop is not provided", async () => {
    const src = "https://cdn.arcnight.com/good";
    const fallback = "F";

    render(<Avatar src={src} fallback={fallback} data-testid="avatar" />);

    await waitFor(() => {
      const avatarProvider = screen.getByTestId("avatar-provider");
      expect(avatarProvider).not.toHaveStyle("--avatar-size: small");
      expect(avatarProvider).not.toHaveStyle("--avatar-size: large");
    });
  });

  it("renders with provided size", async () => {
    const src = "https://cdn.arcnight.com/good";
    const fallback = "F";
    const size = "large";

    render(
      <Avatar src={src} fallback={fallback} size={size} data-testid="avatar" />,
    );

    await waitFor(() => {
      const avatarProvider = screen.getByTestId("avatar-provider");
      expect(avatarProvider).toHaveClass(`avatar-${size}`);
    });
  });

  it("renders with default color if fallback is not provided", async () => {
    const src = "https://cdn.arcnight.com/good";

    render(<Avatar src={src} data-testid="avatar" />);

    await waitFor(() => {
      const avatarProvider = screen.getByTestId("avatar-provider");
      expect(avatarProvider).toHaveStyle("--avatar-color: var(--primary)");
    });
  });

  it("renders with generated color based on fallback", async () => {
    const src = "";
    const fallback = "F";

    render(<Avatar src={src} fallback={fallback} data-testid="avatar" />);

    await waitFor(() => {
      const avatarProvider = screen.getByTestId("avatar-provider");
      expect(avatarProvider).toHaveStyle(`--avatar-color: #347485`);
    });
  });
});
