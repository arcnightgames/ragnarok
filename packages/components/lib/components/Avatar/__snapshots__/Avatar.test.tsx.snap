// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar Component > renders correctly 1`] = `
<div>
  <button
    class="avatar-container rounded-full"
    style="--avatar-color: #347485;"
  />
</div>
`;

exports[`Avatar Component > renders with provided bad src 1`] = `
<div>
  <button
    class="avatar-container rounded-full"
    style="--avatar-color: #347485;"
  />
</div>
`;
